components/Tabs/src/TabBarView/MDCTabBarView.m

```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
+    CGSize contentSize = itemView.intrinsicContentSize;
+    if (contentSize.width > maxWidth) {
+      maxWidth = contentSize.width;
+    }
   }
-  CGSize boundsSize = CGSizeMake(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds));
-  CGSize itemSize = CGSizeMake(boundsSize.width / self.items.count, boundsSize.height);
+  CGFloat requiredWidth = maxWidth * self.items.count;
+  BOOL canBeJustified = availableWidth > requiredWidth;
+  self.containerView.distribution = canBeJustified ? UIStackViewDistributionFillEqually
+                                                   : UIStackViewDistributionFillProportionally;
+}
 
-  CGPoint itemOrigin = CGPointMake(CGRectGetMinX(self.bounds), CGRectGetMinY(self.bounds));
-  for (UIView *itemView in self.itemViews) {
-    itemView.frame = CGRectMake(itemOrigin.x, itemOrigin.y, itemSize.width, itemSize.height);
-    itemOrigin = CGPointMake(itemOrigin.x + itemSize.width, itemOrigin.y);
+- (void)updateConstraints {
+  [super updateConstraints];
+
+  if (self.viewConstraints) {
+    return;
   }
+
+  NSMutableArray *constraints = [NSMutableArray array];
+  [constraints addObjectsFromArray:@[
+    [_containerView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
+    [_containerView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
+    [_containerView.widthAnchor constraintGreaterThanOrEqualToAnchor:self.widthAnchor],
+    [_containerView.topAnchor constraintEqualToAnchor:self.topAnchor],
+    [_containerView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
+  ]];
+
+  [self addConstraints:constraints];
+  self.viewConstraints = constraints;
 }
 
 - (CGSize)intrinsicContentSize {
-  return CGSizeMake(UIViewNoIntrinsicMetric, kMinHeight);
+  CGFloat totalWidth = 0;
+  CGFloat maxHeight = 0;
+  for (UIView *itemView in self.containerView.arrangedSubviews) {
+    CGSize contentSize = itemView.intrinsicContentSize;
+    totalWidth += contentSize.width;
+    if (contentSize.height > maxHeight) {
+      maxHeight = contentSize.height;
+    }
+  }
+  return CGSizeMake(totalWidth, MAX(kMinHeight, maxHeight));
 }
 
 - (CGSize)sizeThatFits:(CGSize)size {
-  return CGSizeMake(size.width, MAX(size.height, kMinHeight));
+  CGSize intrinsicSize = self.intrinsicContentSize;
+  return CGSizeMake(MIN(intrinsicSize.width, size.width), MAX(intrinsicSize.height, size.height));
+}
+
+#pragma mark - Private Functions
+
+- (void)commonMDCTabBarViewInit {
+  self.backgroundColor = UIColor.whiteColor;
+
+  [self setUpStackView];
+  [self setUpItemViews];
+}
+
+- (void)setUpStackView {
+  _containerView = [[UIStackView alloc] init];
+  _containerView.axis = UILayoutConstraintAxisHorizontal;
+  _containerView.translatesAutoresizingMaskIntoConstraints = NO;
+  [self addSubview:_containerView];
+}
+
+- (void)setUpItemViews {
+  for (UIView *view in self.containerView.arrangedSubviews) {
+    [view removeFromSuperview];
+    [_containerView removeArrangedSubview:view];
```

Comment:
This isn't necessary since `removeFromSuperview` automatically removes it from the stack view's `arrangedSubviews` property.

Reply:
Removed.